[R2] Add extensions and unicode interop documentation (#4330)
cloudflare/cloudflare-docs: content/r2/platform/unicode-interoperability.md

---
title: Filename encoding and interoperability problems
pcx-content-type: reference
---

## Filename encoding and interoperability problems

Since R2 is built on top of Workers, it supports Unicode natively. One nuance of Unicode that is often overlooked is the issue of [filename interoperability](https://en.wikipedia.org/wiki/Filename#Encoding_indication_interoperability) due to [Unicode equivalence](https://en.wikipedia.org/wiki/Unicode_equivalence).

> Based on feedback from our users, we've chosen to NFC-normalize key names before storing by default. This means `Héllo` and `Héllo` are the same object in R2 but different objects in other storage providers.

Reviewer: ```suggestion
Based on feedback from our users, we have chosen to NFC-normalize key names before storing by default. This means `Héllo` and `Héllo` are the same object in R2 but different objects in other storage providers.
```

R2 is encoding-preserving for display though. When you list the objects, you'll get back the last encoding you uploaded with.

There are still some platform-specific differences to consider:

* Windows filenames are case-insensitive while R2, Linux, and macOS are not.
* Windows console support for Unicode can be error-prone. Make sure to run `chcp 65001` before using command-line tools or use Cygwin if your object names appear to be incorrect.
* Linux allows distinct files that are unicode-equivalent because filenames are byte streams. Unicode-equivalent filenames on Linux will point to the same R2 object.

> If it's important for you to be able to bypass the unicode equivalence and use byte-oriented key names, contact your Cloudflare account team.

Reviewer: ```suggestion
If it is important for you to be able to bypass the unicode equivalence and use byte-oriented key names, contact your Cloudflare account team.
```

---
title: Filename encoding and interoperability problems
pcx-content-type: reference
---

## Filename encoding and interoperability problems

> Since R2 is built on top of Workers, it supports Unicode natively. One nuance of Unicode that is often overlooked is the issue of [filename interoperability](https://en.wikipedia.org/wiki/Filename#Encoding_indication_interoperability) due to [Unicode
> equivalence](https://en.wikipedia.org/wiki/Unicode_equivalence).

Reviewer 2: ```suggestion
Since R2 is built on top of Workers, it supports Unicode natively. One nuance of Unicode that is often overlooked is the issue of [filename interoperability](https://en.wikipedia.org/wiki/Filename#Encoding_indication_interoperability) due to [Unicode equivalence](https://en.wikipedia.org/wiki/Unicode_equivalence).
```

---
title: Filename encoding and interoperability problems
pcx-content-type: reference

> meta:
>   title: Filename encoding and interoperability problems

Reviewer 2: ```suggestion
```

---

## Filename encoding and interoperability problems

> Since R2 is built on top of Workers, it supports Unicode natively. One nuance of Unicode that is often overlooked
> is the issue of [filename interoperability](https://en.wikipedia.org/wiki/Filename#Encoding_indication_interoperability) due to [Unicode
> equivalence](https://en.wikipedia.org/wiki/Unicode_equivalence).

Reviewer 2: ```suggestion
Since R2 is built on top of Workers, it supports Unicode natively. One nuance of Unicode that is often overlooked is the issue of [filename interoperability](https://en.wikipedia.org/wiki/Filename#Encoding_indication_interoperability) due to [Unicode
equivalence](https://en.wikipedia.org/wiki/Unicode_equivalence).
```

> Based on feedback from our users, we've chosen to NFC-normalize key names before storing by default. This means
> `Héllo` and `Héllo` are the same object in R2 but different objects in other storage providers.

Reviewer 2: @vlovich Can you provide the byte codes so that the two `Héllo` are noticeably different?

```suggestion
Based on user feedback, Cloudflare will NFC-normalize key names before storing by default. This means `Héllo` and `Héllo` are the same object in R2 but different objects in other storage providers.
```

> R2 is encoding-preserving for display though. When you list the objects, you'll get back the last encoding you uploaded with.

Reviewer 2: Not sure what you mean by "for display though" @vlovich 

```suggestion
R2 preserves the encoding for display though. When you list the objects, you'll get back the last encoding you uploaded with.
```

> There are still some platform-specific differences to consider
> 
> * Windows filenames are case-insensitive while R2, Linux, and macOS are not.
> * Windows console support for Unicode can be error-prone. Make sure to run `chcp 65001` before using command-line tools or use Cygwin
> if your object names are coming out garbled.
> * Linux allows distinct files that are unicode equivalent because filenames are byte streams. Unicode equivalent files on Linux
> will map to the same R2 object.
> 
> If it's important for you to be able to bypass the unicode equivalence and use byte-oriented key names, contact your Cloudflare account
> team.

Reviewer 2: ```suggestion
There are still some platform-specific differences to consider:

* Windows filenames are case-insensitive while R2, Linux, and macOS are not.
* Windows console support for Unicode can be error-prone. Make sure to run `chcp 65001` before using command-line tools or use Cygwin if your object names appear to be incorrect.
* Linux allows distinct files that are unicode-equivalent because filenames are byte streams. Unicode-equivalent filenames on Linux will point to the same R2 object.

If it's important for you to be able to bypass the unicode equivalence and use byte-oriented key names, contact your Cloudflare account team.
```
